### PR TITLE
Add support for TLS/SSL mutual authentication

### PR DESCRIPTION
**Pull request by bplost**
Add support for client certificat authentication in TLS_FTPHandler

### CODE REVIEW

File: pyftpdlib/handlers.py [hunk 1]
```diff
@@ -3414,6 +3415,9 @@ class TLS_FTPHandler(SSLConnection, FTPHandler):
         certfile = None
         keyfile = None
         ssl_protocol = SSL.SSLv23_METHOD
+		# client certificate configurable attributes
```
Comment (lurch) [inline pyftpdlib/handlers.py]: Looks like this line is still using tabs, rather than spaces.

**Comment by lurch** [comment]
Just out of curiosity, what happens when the certificate is invalid? Does it *just* print "Bad client certificate detected" or does the connection get refused? Should some kind of warning or exception be raised, so that it can be detected programmatically?

**Comment by bplost** [comment]
Connection gets rejected.  The error sent to the client through pyOpenSSL is a standard ssl error with the correct information in it - the client I'm using (WinSCP) shows that the certificate is invalid (or if no certificate is sent, then it shows that the server is expecting a certificate).  Ia'll add screenshots when i create a new pull request.

**Comment by lurch** [comment]
No need to keep creating new PRs - you can simply update the existing PR with `git push -f`

**Comment by bplost** [comment]
Thanks @lurch, will do

**Comment by bplost** [comment]
Client error screenshots (WinSCP) when attempting connection with a bad certificate and with no certificate.
nocertificate - clienterror
badcertificate - clienterror

**Comment by bplost** [comment]
Anyone have ideas why Python 3.x is getting socket timeout errors in the AppVeyor build/test?

File: pyftpdlib/handlers.py [hunk 2]
```diff
@@ -3449,6 +3453,15 @@ def __init__(self, conn, server, ioloop=None):
 
         def __repr__(self):
             return FTPHandler.__repr__(self)
+
```
Comment (lurch) [inline pyftpdlib/handlers.py]: `flake8` (which is run by the Travis builds) is complaining about this line being blank, but still containing space characters

File: pyftpdlib/handlers.py [hunk 3]
```diff
@@ -22,6 +22,7 @@
 
 try:
     from OpenSSL import SSL  # requires "pip install pyopenssl"
+    from OpenSSL.SSL import  VERIFY_PEER, VERIFY_FAIL_IF_NO_PEER_CERT, VERIFY_CLIENT_ONCE, SESS_CACHE_OFF, OP_NO_TICKET
```
Comment (lurch) [inline pyftpdlib/handlers.py]: `flake8` (which is run by the Travis builds) is complaining about there being two spaces after `import`

File: pyftpdlib/handlers.py [hunk 4]
```diff
@@ -3449,6 +3453,15 @@ def __init__(self, conn, server, ioloop=None):
 
         def __repr__(self):
             return FTPHandler.__repr__(self)
+            
+        #commented out prints - for different Python version support
```
Comment (lurch) [inline pyftpdlib/handlers.py]: `flake8` (which is run by the Travis builds) is complaining about this line (and the others below) not having a space after the `#` character

File: pyftpdlib/handlers.py [hunk 5]
```diff
@@ -3464,6 +3477,11 @@ def get_ssl_context(cls):
                 if not cls.keyfile:
                     cls.keyfile = cls.certfile
                 cls.ssl_context.use_privatekey_file(cls.keyfile)
+                if cls.tls_mutual_authentication:
+                    cls.ssl_context.set_verify(VERIFY_PEER | VERIFY_FAIL_IF_NO_PEER_CERT | VERIFY_CLIENT_ONCE, cls.verify_certs_callback)
```
Comment (lurch) [inline pyftpdlib/handlers.py]: `flake8` is also complaining about this line being too long, but I dunno what @giampaolo 's preferred line-wrapping style is ;-)

**Comment by giampaolo** [comment]
Some preliminary questions:

- how does this relate to `certfile` and `keyfile` class attributes? Is `tls_mutual_certfile` supposed to be different than `certfile` or they can be the same?
- what happens exactly in case the client does not provide a certificate? I see a callback function is called but what will the server do? Raise an exception? At the very least the callback function should log the event.

Some considerations:

- the `tls_mutual_authentication` bool option looks unnecessary; I guess you can simply ad/define a `tls_mutual_certfile` option (which I would rename to `mutual_certfile`)
- the PR is lacking tests

File: pyftpdlib/handlers.py (hunk 3 above)
Comment (giampaolo) [inline pyftpdlib/handlers.py]: there should be one import per line; also imports should be sorted alphabetically

File: pyftpdlib/handlers.py [hunk 6]
```diff
@@ -3449,6 +3453,15 @@ def __init__(self, conn, server, ioloop=None):
...
             return FTPHandler.__repr__(self)
+            
+        #commented out prints - for different Python version support
+        @staticmethod
+        def verify_certs_callback(connection, x509, errnum, errdepth, ok):
+            #if not ok:
+                #print "Bad client certificate detected."
+            #else:
+                #print "Client certificate ok."
+            return ok
```
Comment (giampaolo) [inline pyftpdlib/handlers.py]: If this is called when a client connects, as I suppose, then it should be a `classmethod`.

**Comment by bplost** [comment]
Answers to your questions @giampaolo 

 - The mutual certfile is what the server checks the client's certificate against, to verify that the client is allowed to connect to that particular server.  It _should_ be different than certfile and keyfile
 - When the client provides no certificate, the OpenSSL library takes care of sending the correct response to the client ([see screenshot above](https://cloud.githubusercontent.com/assets/18441945/17899765/069ae4f2-6922-11e6-9a94-71f595955954.png)).  I will add logging here.
 - Agreed that the bool option is unnecessary and will remove
 - Will add tests

**Comment by giampaolo** [comment]
> When the client provides no certificate, the OpenSSL library takes care of sending the correct 
> response to the client (see screenshot above). I will add logging here.

But what happens on the server side? The client gets disconnected? When exactly? Is an exception raised? Is the callback function called before or after that happens?
Ideally, the server should gracefully handle that event, meaning catch the exception and log a message. If OpenSSL takes care of sending "something" which tells the client it should provide a certificate, then fine.

**Comment by bplost** [comment]
The client gets disconnected during the SSL handshake.  Sample log:
noclientcertlog

After the handshake fails, the checks you already have in place take care of handling it gracefully and the server just keeps going.
The callback function is _not_ called when no certificate is passed (as the failure happens earlier).  The callback function is called when a client certificate is passed (valid or invalid).  If the certificate is valid, the connection is allowed.  If the certificate is invalid, again the OpenSSL library sends the correct message to the client that the cert is invalid, and I will add the internal logging into the callback method.

**Comment by bplost** [comment]
Cleaned up code and added tests

File: pyftpdlib/test/test_functional_ssl_client_certfile.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+FTPS_SUPPORT = hasattr(ftplib, 'FTP_TLS')
+if sys.version_info < (2, 7):
+    FTPS_UNSUPPORT_REASON = "requires python 2.7+"
+else:
+    FTPS_UNSUPPORT_REASON = "FTPS test skipped"
+
+CERTFILE = os.path.abspath(os.path.join(os.path.dirname(__file__),
+                                        'keycert.pem'))
+CLIENT_CERTFILE = os.path.abspath(os.path.join(os.path.dirname(__file__),
+                                        'clientcert.pem'))
```
Comment (lurch) [inline pyftpdlib/test/test_functional_ssl_client_certfile.py]: Looks like you forgot to add this `clientcert.pem` file to the PR?

**Comment by lurch** [comment]
@giampaolo Apparently flake8 no longer supports Python 2.6 (https://gitlab.com/pycqa/flake8/issues/187 )
Could the Travis config be modified to not run flake8 on python2.6 ? (which would then stop Travis errors like https://travis-ci.org/giampaolo/pyftpdlib/jobs/155328753 )

**Comment by bplost** [comment]
I think at this point all of the failing tests are not due to my changes

**Comment by zuohaocheng** [comment]
This PR is refined by #428 and can be closed.

**Pull request closed without merging**